relay-general/src/store/normalize/spans.rs

```diff
@@ -112,33 +112,82 @@ pub fn normalize_spans(event: &mut Event, attributes: &BTreeSet<SpanAttribute>)
-
-    let mut span_map = HashMap::new();
+fn set_event_exclusive_time(
+    event: &mut Event,
+    span_map: &mut HashMap<SpanId, Vec<TimeWindowSpan>>,
+) {
+    let start = match event.start_timestamp.value() {
+        Some(timestamp) => *timestamp,
+        _ => return,
+    };
+
+    let end = match event.timestamp.value() {
+        Some(timestamp) => *timestamp,
+        _ => return,
+    };
+
+    let span_interval = TimeWindowSpan::new(start, end);
+
+    let contexts = match event.contexts.value_mut() {
+        Some(contexts) => contexts,
+        _ => return,
+    };
+
+    let trace_context = match contexts.get_mut("trace").map(Annotated::value_mut) {
+        Some(Some(ContextInner(Context::Trace(trace_context)))) => trace_context,
+        _ => return,
+    };
+
+    let span_id = match trace_context.span_id.value() {
+        Some(span_id) => span_id,
+        _ => return,
+    };
+
+    let child_intervals = match span_map.get_mut(span_id) {
+        Some(intervals) => {
+            // Make sure that the intervals are sorted by start time.
+            intervals.sort_unstable_by_key(|interval| interval.start);
+            merge_non_overlapping_intervals(intervals)
+        }
+        None => Vec::new(),
+    };
 
-    for span in spans.iter() {
-        let span = match span.value() {
-            None => continue,
-            Some(span) => span,
-        };
+    let exclusive_time = interval_exclusive_time(&span_interval, &child_intervals);
 
-        let parent_span_id = match span.parent_span_id.value() {
-            None => continue,
-            Some(parent_span_id) => parent_span_id.clone(),
-        };
+    trace_context.exclusive_time = Annotated::new(exclusive_time);
+}
 
-        let interval = match get_span_interval(span) {
-            None => continue,
-            Some(interval) => interval,
-        };
+fn compute_span_exclusive_time(event: &mut Event) {
+    let mut span_map = HashMap::new();
 
-        span_map
-            .entry(parent_span_id)
-            .or_insert_with(Vec::new)
-            .push(interval)
+    if let Some(spans) = event.spans.value() {
+        for span in spans.iter() {
+            let span = match span.value() {
+                None => continue,
+                Some(span) => span,
+            };
+
+            let parent_span_id = match span.parent_span_id.value() {
+                None => continue,
+                Some(parent_span_id) => parent_span_id.clone(),
+            };
+
+            let interval = match get_span_interval(span) {
+                None => continue,
+                Some(interval) => interval,
+            };
+
+            span_map
+                .entry(parent_span_id)
+                .or_insert_with(Vec::new)
+                .push(interval)
+        }
     }
 
+    set_event_exclusive_time(event, &mut span_map);
+
+    let spans = event.spans.value_mut().get_or_insert_with(|| Vec::new());
```

Comment:
I notice this inserts a new vector if `event.spans` doesn't have one set. @untitaker would this be fine?

Reply:
apparently we already did that before so I am not too worried about that

you can avoid the double-borrowing here if you change `set_event_exclusive_time` to take timestamps and a mutable reference to contexts instead of the entire event. then you can still hold a mutable reference to spans throughout this entire function